src/core/geocms/qgsgeonoderequest.cpp

```diff
@@ -0,0 +1,405 @@
+      if ( layerTypeName.length() == 0 )
+      {
+        layerTypeName = layer["distribution_url"].toString().split( "/" ).last();
+      }
+      // On this step, layerTypeName is in WORKSPACE%3ALAYERNAME or WORKSPACE:LAYERNAME format
+      if ( layerTypeName.contains( "%3A" ) )
+      {
+        layerTypeName.replace( "%3A", ":" );
+      }
+      // On this step, layerTypeName is in WORKSPACE:LAYERNAME format
+      QStringList splitURL = layerTypeName.split( ":" );
+      QString layerWorkspace = splitURL[0];
+      QString layerName = splitURL[1];
+
+      layerStruct.name = layerName;
+      layerStruct.typeName = layerTypeName;
+      layerStruct.uuid = layer["uuid"].toString();
+      layerStruct.title = layer["title"].toString();;
+
+      // WMS url : BASE_URI/geoserver/WORKSPACE/wms
+      layerStruct.wmsURL = mBaseUrl + "/geoserver/" + layerWorkspace + "/wms";
+      // WFS url : BASE_URI/geoserver/WORKSPACE/wfs
+      layerStruct.wfsURL = mBaseUrl + "/geoserver/" + layerWorkspace + "/wfs";
+      // XYZ url : set to empty string
+      layerStruct.xyzURL = "";
+
+      layers.append( layerStruct );
+    }
+  }
+  // Geonode version 2.7 or newer
+  else if ( ( majorVersion == 2 && minorVersion >= 7 ) || ( majorVersion >= 3 ) )
+  {
+    for ( int i = 0; i < layerList.count(); i++ )
+    {
+      LayerStruct layerStruct;
+      QVariantMap layer = layerList[i].toMap();
+      // Find WMS, WFS, and XYZ link
+      QVariantList layerLinks = layer["links"].toList();
+      layerStruct.wmsURL = QStringLiteral( "" );
+      layerStruct.wfsURL = QStringLiteral( "" );
+      layerStruct.xyzURL = QStringLiteral( "" );
+      for ( int j = 0; j < layerLinks.count(); j++ )
+      {
+        QVariantMap link = layerLinks[j].toMap();
+        if ( link.contains( "link_type" ) )
+        {
+          if ( link["link_type"] == "OGC:WMS" )
+          {
+            layerStruct.wmsURL = link["url"].toString();
+          }
+          if ( link["link_type"] == "OGC:WFS" )
+          {
+            layerStruct.wfsURL = link["url"].toString();
+          }
+          if ( link["link_type"] == "image" )
+          {
+            if ( link.contains( "name" ) && link["name"] == "Tiles" )
+            {
+              layerStruct.xyzURL = link["url"].toString();
+            }
+          }
+        }
+      }
+      if ( layer["typename"].toString().length() == 0 )
+      {
+        QStringList splitURL = layer["detail_url"].toString().split( "/" );
+        layerStruct.typeName = splitURL[ splitURL.length() - 1];
+      }
+      layerStruct.uuid = layer["uuid"].toString();
+      layerStruct.name = layer["name"].toString();
+      layerStruct.typeName = layer["typename"].toString();
+      layerStruct.title = layer["title"].toString();
+      layers.append( layerStruct );
+    }
+  }
+  return layers;
+}
+
+
+QStringList QgsGeoNodeRequest::serviceUrls( QString serviceType )
+{
+  QStringList *urls = new QStringList;
```

Comment:
Do not use ```QStringList *urls = new QStringList``` here, it will leak memory... just do ```QStringList urls;``` - simpler and does not leak :-)